File: ui-service/src/lib/intake/indiaStateIdentity.ts
/** Canonical India state identity — names, geo-master codes, and ISO 3166-2:IN short codes. */

const SHORT_CODE_TO_TOKEN: Record<string, string> = {
  AN: 'ANDAMANANDNICOBARISLANDS',
  AP: 'ANDHRAPRADESH',
  AR: 'ARUNACHALPRADESH',
  AS: 'ASSAM',
  BR: 'BIHAR',
  CH: 'CHANDIGARH',
  CT: 'CHHATTISGARH',
  CG: 'CHHATTISGARH',
  DN: 'DADRAANDNAGARHAVELI',
  DD: 'DAMANANDDIU',
  DL: 'DELHI',
  GA: 'GOA',
  GJ: 'GUJARAT',
  HR: 'HARYANA',
  HP: 'HIMACHALPRADESH',
  JK: 'JAMMUANDKASHMIR',
  JH: 'JHARKHAND',
  KA: 'KARNATAKA',
  KL: 'KERALA',
  LA: 'LADAKH',
  LD: 'LAKSHADWEEP',
  MP: 'MADHYAPRADESH',
  MH: 'MAHARASHTRA',
  MN: 'MANIPUR',
  ML: 'MEGHALAYA',
  MZ: 'MIZORAM',
  NL: 'NAGALAND',
  OR: 'ODISHA',
  OD: 'ODISHA',
  PY: 'PUDUCHERRY',
  PB: 'PUNJAB',
  RJ: 'RAJASTHAN',
  SK: 'SIKKIM',
  TN: 'TAMILNADU',
  TG: 'TELANGANA',
  TS: 'TELANGANA',
  TR: 'TRIPURA',
  UP: 'UTTARPRADESH',
  UT: 'UTTARAKHAND',
  UK: 'UTTARAKHAND',
  WB: 'WESTBENGAL',
}

export function indiaStateIdentity(raw: string | null | undefined): string {
  const token = String(raw ?? '')
    .toUpperCase()
    .replace(/[^A-Z0-9]/g, '')
  if (!token) return ''
  return SHORT_CODE_TO_TOKEN[token] ?? token
}

export function indiaStateMatchesAllowed(candidate: string, allowed: string[] | null | undefined): boolean {
  if (!allowed || allowed.length === 0) return true
  const identity = indiaStateIdentity(candidate)
  if (!identity) return true
  return allowed.some((a) => indiaStateIdentity(a) === identity)
}

File: los-core-service/src/main/java/com/los/core/service/master/IndiaStateIdentity.java
package com.los.core.service.master;

import java.util.Collection;
import java.util.Locale;
import java.util.Map;

/**
 * Canonical India state identity for workflow {@code allowedStates} matching.
 *
 * <p>Workflow admin stores geo-master display names (e.g. {@code Tamil Nadu}).
 * Intake APIs may persist ISO-style codes ({@code TN}) or geo-master codes
 * ({@code TAMIL_NADU}). All three must match the same identity.
 */
public final class IndiaStateIdentity {

    /**
     * ISO 3166-2:IN two-letter codes (and common two-letter aliases) → alphanumeric token.
     */
    private static final Map<String, String> SHORT_CODE_TO_TOKEN = Map.ofEntries(
            Map.entry("AN", "ANDAMANANDNICOBARISLANDS"),
            Map.entry("AP", "ANDHRAPRADESH"),
            Map.entry("AR", "ARUNACHALPRADESH"),
            Map.entry("AS", "ASSAM"),
            Map.entry("BR", "BIHAR"),
            Map.entry("CH", "CHANDIGARH"),
            Map.entry("CT", "CHHATTISGARH"),
            Map.entry("CG", "CHHATTISGARH"),
            Map.entry("DN", "DADRAANDNAGARHAVELI"),
            Map.entry("DD", "DAMANANDDIU"),
            Map.entry("DL", "DELHI"),
            Map.entry("GA", "GOA"),
            Map.entry("GJ", "GUJARAT"),
            Map.entry("HR", "HARYANA"),
            Map.entry("HP", "HIMACHALPRADESH"),
            Map.entry("JK", "JAMMUANDKASHMIR"),
            Map.entry("JH", "JHARKHAND"),
            Map.entry("KA", "KARNATAKA"),
            Map.entry("KL", "KERALA"),
            Map.entry("LA", "LADAKH"),
            Map.entry("LD", "LAKSHADWEEP"),
            Map.entry("MP", "MADHYAPRADESH"),
            Map.entry("MH", "MAHARASHTRA"),
            Map.entry("MN", "MANIPUR"),
            Map.entry("ML", "MEGHALAYA"),
            Map.entry("MZ", "MIZORAM"),
            Map.entry("NL", "NAGALAND"),
            Map.entry("OR", "ODISHA"),
            Map.entry("OD", "ODISHA"),
            Map.entry("PY", "PUDUCHERRY"),
            Map.entry("PB", "PUNJAB"),
            Map.entry("RJ", "RAJASTHAN"),
            Map.entry("SK", "SIKKIM"),
            Map.entry("TN", "TAMILNADU"),
            Map.entry("TG", "TELANGANA"),
            Map.entry("TS", "TELANGANA"),
            Map.entry("TR", "TRIPURA"),
            Map.entry("UP", "UTTARPRADESH"),
            Map.entry("UT", "UTTARAKHAND"),
            Map.entry("UK", "UTTARAKHAND"),
            Map.entry("WB", "WESTBENGAL")
    );

    private IndiaStateIdentity() {}

    public static boolean matchesAllowed(String candidate, Collection<String> allowed) {
        if (allowed == null || allowed.isEmpty()) {
            return true;
        }
        String identity = identity(candidate);
        if (identity.isEmpty()) {
            return true;
        }
        for (String allowedValue : allowed) {
            if (identity.equals(identity(allowedValue))) {
                return true;
            }
        }
        return false;
    }

    public static String identity(String raw) {
        String token = alphanumeric(raw);
        if (token.isEmpty()) {
            return "";
        }
        String fromShort = SHORT_CODE_TO_TOKEN.get(token);
        if (fromShort != null) {
            return fromShort;
        }
        return token;
    }

    private static String alphanumeric(String raw) {
        if (raw == null) {
            return "";
        }
        StringBuilder sb = new StringBuilder(raw.length());
        for (int i = 0; i < raw.length(); i++) {
            char c = raw.charAt(i);
            if (Character.isLetterOrDigit(c)) {
                sb.append(c);
            }
        }
        return sb.toString().toUpperCase(Locale.ROOT);
    }
}
